fuzzing/reversion/reversion_statistics.go

```diff
@@ -0,0 +1,188 @@
+		}
+
+		// Disregard deployments
+		if msg.To == nil {
+			continue
+		}
+
+		if len(msg.Data) < 4 {
+			continue
+		}
+
+		funcSelector := functionSelector{}
+		copy(funcSelector[:], msg.Data[:4])
+
+		if msgResult.ExecutionResult.Err != nil {
+			revertReason := getRevertReason(msgResult.ExecutionResult)
+			report.addCall(funcSelector, revertReason, true)
+		} else {
+			report.addCall(funcSelector, errorSelector{}, false)
+		}
+	}
+	if len(s.incomingReportsQueue) >= 95 {
+		fmt.Printf("incoming reports queue full")
+	}
+
+	s.incomingReportsQueue <- report
+	return nil
+}
+
+// BuildArtifact converts aggregated report information into an artifact that can be easily serialized.
+func (s *ReversionMeasurer) BuildArtifact(logger *logging.Logger, contractDefs fuzzerTypes.Contracts, corpusDir string) error {
+	if !s.enabled {
+		return nil
+	}
+	artifact, err := CreateRevertReportArtifact(logger, s.aggReport, contractDefs, corpusDir)
+	if err != nil {
+		return err
+	}
+
+	s.reportArtifact = artifact
+	return nil
+}
+
+// WriteReport takes the generated reportArtifact and writes it to the provided dir. Two artifacts are written;
+// a revert statistics json file, and a user-readable html file.
+func (s *ReversionMeasurer) WriteReport(dir string, logger *logging.Logger) error {
+	if !s.enabled {
+		return nil
+	}
+	if s.reportArtifact == nil {
+		return errors.New("report artifact missing")
+	}
+	jsonpath := filepath.Join(dir, "revert_stats.json")
+	err := s.writeReportJson(jsonpath, logger)
+	if err != nil {
+		return err
+	}
+
+	markdownpath := filepath.Join(dir, "revert_stats.html")
+	err = s.writeReportHtml(markdownpath, logger)
+
+	return err
+}
+
+// writeReportHtml generates an HTML representation of the report artifact and writes it to path.
+func (s *ReversionMeasurer) writeReportHtml(path string, logger *logging.Logger) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	err = s.reportArtifact.ConvertToHtml(file)
+
+	if err == nil {
+		logger.Info("Revert stats report written to: ", path)
+	}
+	return err
+}
+
+// writeReportHtml generates an JSON representation of the report artifact and writes it to path.
+func (s *ReversionMeasurer) writeReportJson(path string, logger *logging.Logger) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	b, err := json.MarshalIndent(s.reportArtifact, "", "    ")
+	if err != nil {
+		return err
+	}
+	_, err = file.Write(b)
+	if err == nil {
+		logger.Info("Revert stats written to file: ", path)
+	}
+	return err
+}
+
+// getRevertReason encodes the error from `result` into an errorSelector. If the error is a solidity-inserted revert,
+// the returnData is decoded to identify the type of revert, and this id is smuggled into the last byte of errorSelector.
+// These smuggled errors can be decoded using decodeSmuggledSolidityRevertReason.
+func getRevertReason(result *core.ExecutionResult) errorSelector {
```

Comment:
Let's use
https://github.com/crytic/medusa/blob/3a9b0fae7888b1c1972c1178ca5022ebd51f5414/fuzzing/executiontracer/execution_trace.go#L218

Reply:
this code does use GetSolidityPanicCode further down. we pack panicCode into the last byte of the error selector so we don't have to create a special data type for all the maps/counters to key on